src/librustc/ty/instance.rs

```diff
@@ -187,45 +187,7 @@ impl<'tcx> Instance<'tcx> {
         def_id: DefId,
         substs: SubstsRef<'tcx>,
     ) -> Option<Instance<'tcx>> {
-        debug!("resolve(def_id={:?}, substs={:?})", def_id, substs);
-        let result = if let Some(trait_def_id) = tcx.trait_of_item(def_id) {
-            debug!(" => associated item, attempting to find impl in param_env {:#?}", param_env);
-            let item = tcx.associated_item(def_id);
-            resolve_associated_item(tcx, &item, param_env, trait_def_id, substs)
-        } else {
-            let ty = tcx.type_of(def_id);
-            let item_type = tcx.subst_and_normalize_erasing_regions(substs, param_env, &ty);
-
-            let def = match item_type.kind {
-                ty::FnDef(..)
-                    if {
-                        let f = item_type.fn_sig(tcx);
-                        f.abi() == Abi::RustIntrinsic || f.abi() == Abi::PlatformIntrinsic
-                    } =>
-                {
-                    debug!(" => intrinsic");
-                    ty::InstanceDef::Intrinsic(def_id)
-                }
-                _ => {
-                    if Some(def_id) == tcx.lang_items().drop_in_place_fn() {
-                        let ty = substs.type_at(0);
-                        if ty.needs_drop(tcx, ty::ParamEnv::reveal_all()) {
-                            debug!(" => nontrivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, Some(ty))
-                        } else {
-                            debug!(" => trivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, None)
-                        }
-                    } else {
-                        debug!(" => free item");
-                        ty::InstanceDef::Item(def_id)
-                    }
-                }
-            };
-            Some(Instance { def: def, substs: substs })
-        };
-        debug!("resolve(def_id={:?}, substs={:?}) = {:?}", def_id, substs, result);
-        result
+        tcx.resolve_instance((param_env, def_id, substs))
```

Comment:
Hmm, this seems like a place where canonicalization would be appropriate. We don't expect inference variables, but we *do* use this (I believe) without having fully erased regions, right? In that case, canonicalizing might eliminate duplicate work.

Reply:
I agree. However, I think https://github.com/rust-lang/rust/pull/67717 is already doing canonicalization changes related to (among other things) `Instance::resolve`.

Reply:
Heh, I just came here from reviewing #67717 to point out that it is making changes to `Instance::resolve` that would require canonicalization.

Reply:
@nikomatsakis: Would you like me to add canonicalization in this PR, or should it be done in a follow-up PR after #67717 is merged?